Web Inspector: Add a toggle to hide UA stylesheet rules in Elements tab

J0pan opened the pull request:
#### 600cfea07b287a49b35bc6ed568c6683a1699a73
<pre>
Web Inspector: Add a toggle to hide UA stylesheet rules in Elements tab
<a href="https://bugs.webkit.org/show_bug.cgi?id=302161">https://bugs.webkit.org/show_bug.cgi?id=302161</a>
<a href="https://rdar.apple.com/164265183">rdar://164265183</a>

Reviewed by NOBODY.

This change adds a setting in the Setting &gt; Elements tab to control whether User Agent styles are displayed.
When disabled, User Agent rules are filtered out at section creation time in the Elements &gt; Styles panel.
Under the computed Panel the &quot;go-to&quot; arrows situated beside each style property are also hidden if
the style property points to a User Agent Style and the &quot;show user agent styles&quot; settings is disabled.

No new tests.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
(WI.ComputedStyleDetailsPanel.prototype.initialLayout):
(WI.ComputedStyleDetailsPanel.prototype._handleShowUserAgentStylesChanged):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.layout):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._shouldIncludeStyle):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._handleShowUserAgentStylesChanged):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype.update):

Canonical link: <a href="https://commits.webkit.org/304627@main">https://commits.webkit.org/304627@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/898bd98cd58d98ac4035857b54277356847f560a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136106 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143812 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104051 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6662 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121972 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84897 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3964 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4408 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40165 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146560 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8146 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40733 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112404 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112756 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28617 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6224 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118272 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8194 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36335 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7910 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71753 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8133 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7986 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -300,9 +302,16 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
                 else if (pseudoId === WI.CSSManager.PseudoSelectorNames.After)
                     pseudoElement = this.nodeStyles.node.afterPseudoElement();
+
+                let orderedPseudoStyles = WI.DOMNodeStyles.uniqueOrderedStyles(pseudoElementInfo.orderedStyles).filter((style) => this._shouldIncludeStyle(style));
```
if some of these styles are hidden, how will that affect other features that try to jump to it?

for example, viewing the "trace" of a particular property in the Computed panel will probably show a go-to arrow for these styles that no longer exist, so i think we may need some logic there as well (and maybe other places im not thinking of) to at the very least hide the go-to arrow

J0pan replied:
It now hides the arrows that trace to a UA style. 

I thought of another way such that we would untoggle and "turn on" show user agent styles when a user clicks on a computed value; however, I'm wasn't sure if/when we would retoggle it back to hide UA styles again, it felt like it would be inconsistent for the user.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -300,9 +302,16 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
                 else if (pseudoId === WI.CSSManager.PseudoSelectorNames.After)
                     pseudoElement = this.nodeStyles.node.afterPseudoElement();
+
+                let orderedPseudoStyles = WI.DOMNodeStyles.uniqueOrderedStyles(pseudoElementInfo.orderedStyles).filter((style) => this._shouldIncludeStyle(style));
```
id prefer if there was a way to avoid using `filter` as that has to iterate the entire input array and allocate an entirely new output array, so it's pretty wasteful

but i suppose it's not _that_ expensive since the list of styles is usually in the hundreds or thousands, so it's probably fine (at least until if it's not)

J0pan replied:
Done. It's now using a loop with early exit, it also avoids the extra array allocation.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -358,6 +371,21 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
+
+    _isUserAgentRuleStyle(style)
+    {
+        return style.type === WI.CSSStyleDeclaration.Type.Rule && style.ownerRule && style.ownerRule.type === WI.CSSStyleSheet.Type.UserAgent
```
NIT: we can make this slightly simpler using optional chaining
```js
return style.type === WI.CSSStyleDeclaration.Type.Rule && style.ownerRule?.type === WI.CSSStyleSheet.Type.UserAgent;
```

J0pan replied:
Done

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -358,6 +371,21 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
+
+    _handleShowUserAgentStylesChanged(event)
+    {
+        this.refresh(true);
```
Style: we normally use a variable for literals like this so that it's easier to understand at the callsite what it's for
```js
const significantChange = true;
this.refresh(significantChange);
```

J0pan replied:
Done

dcrousso reviewed (commented):
ooo neat idea!

stwrt commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -358,6 +371,21 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
+
+    _isUserAgentRuleStyle(style)
+    {
+        return style.type === WI.CSSStyleDeclaration.Type.Rule && style.ownerRule && style.ownerRule.type === WI.CSSStyleSheet.Type.UserAgent
```
Missing semi colon at end.

J0pan replied:
Done

stwrt commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -306,7 +306,8 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
         cssGroup.addSetting(WI.settings.showCSSPropertySyntaxInDocumentationPopover, WI.UIString("Show property syntax in documentation popover"));
-
+        cssGroup.addSetting(WI.settings.showUserAgentStyles, WI.UIString("Show user agent styles"));
+        
```
Extra white space.

J0pan replied:
Done

stwrt reviewed (commented):
Style comments

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -300,10 +302,29 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
 
-                for (let style of WI.DOMNodeStyles.uniqueOrderedStyles(pseudoElementInfo.orderedStyles))
+                for (let style of orderedPseudoStyles) {
+                    if (!this._shouldIncludeStyle(style))
```
having to check this twice is i think not as easy to read as when it was all in a single array (even though it did require an extra allocation)

as such, id go back to what you had before
```js
let orderedPseudoStyles = WI.DOMNodeStyles.uniqueOrderedStyles(pseudoElementInfo.orderedStyles)
    .filter((style) => this._shouldIncludeStyle(style));
```

apologies for leading you down the wrong path

dcrousso replied:
another option could be to have `WI.DOMNodeStyles.uniqueOrderedStyles` accept another optional argument that is used as a filter within that function
```diff
-static uniqueOrderedStyles(orderedStyles)
+static uniqueOrderedStyles(orderedStyles, {filter} = {})
 {
     let uniqueOrderedStyles = [];

     for (let style of orderedStyles) {
         let rule = style.ownerRule;
         if (!rule) {
-            uniqueOrderedStyles.push(style);
+            if (!filter || filter(style))
+                uniqueOrderedStyles.push(style);
             continue;
         }

         let found = false;
         for (let existingStyle of uniqueOrderedStyles) {
             if (rule.isEqualTo(existingStyle.ownerRule)) {
                 found = true;
                 break;
             }
         }
-        if (!found)
+        if (!found && (!filter || filter(style)))
             uniqueOrderedStyles.push(style);
     }

     return uniqueOrderedStyles;
 }
```

but honestly we dont have to optimize this prematurely since it's likely only ever gonna be dozens of items at most

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
```diff
@@ -228,31 +228,43 @@ WI.SpreadsheetStyleProperty = class SpreadsheetStyleProperty extends WI.Object
+
+            // Check if it's a user agent style BEFORE potentially replacing with shorthand.
+            // User agent styles don't have styleSheetTextRange, so they get replaced below.
+            let originalOwnerRule = effectiveProperty?.ownerStyle?.ownerRule;
```
i dont think it's possible for `ownerStyle` to not exist
```js
let originalOwnerRule = effectiveProperty?.ownerStyle.ownerRule;
```

if you're seeing otherwise, please file a bug with whatever steps you're using to reproduce as that likely needs to be fixed :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
```diff
@@ -228,31 +228,43 @@ WI.SpreadsheetStyleProperty = class SpreadsheetStyleProperty extends WI.Object
+            if (!shouldShowGoToArrow) {
+                let isUserAgent = originalOwnerRule?.type === WI.CSSStyleSheet.Type.UserAgent;
+                shouldShowGoToArrow = !isUserAgent;
+            }
```
we can inline this
```js
let shouldShowGoToArrow = WI.settings.showUserAgentStyles.value || originalOwnerRule?.type !== WI.CSSStyleSheet.Type.UserAgent;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
```diff
@@ -228,31 +228,43 @@ WI.SpreadsheetStyleProperty = class SpreadsheetStyleProperty extends WI.Object
                 effectiveProperty = effectiveProperty.relatedShorthandProperty;
 
-            let ownerRule = effectiveProperty ? effectiveProperty.ownerStyle.ownerRule : null;
+            let ownerRule = effectiveProperty?.ownerStyle?.ownerRule;
```
ditto about `ownerStyle` always existing
```js
let ownerRule = effectiveProperty?.ownerStyle.ownerRule ?? null;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
```diff
@@ -228,31 +228,43 @@ WI.SpreadsheetStyleProperty = class SpreadsheetStyleProperty extends WI.Object
+                arrowElement.addEventListener("click", (event) => {
+                    if (!effectiveProperty || !ownerRule || !event.altKey) {
+                        if (this._delegate.spreadsheetStylePropertyShowProperty)
+                            this._delegate.spreadsheetStylePropertyShowProperty(this, this._property);
```
while we're making other changes here, let's simplify this
```js
if (!effectiveProperty || !ownerRule || !event.altKey) {
    this._delegate.spreadsheetStylePropertyShowProperty?.(this, this._property);
    return;
}
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -358,6 +383,22 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
+
+    _isUserAgentRuleStyle(style)
+    {
+        return style.type === WI.CSSStyleDeclaration.Type.Rule && style.ownerRule?.type === WI.CSSStyleSheet.Type.UserAgent;
```
does it actually matter if it's `WI.CSSStyleDeclaration.Type.Rule` so long as the `ownerRule` is a `WI.CSSStyleSheet.Type.UserAgent`?
```js
_shouldIncludeStyle(style)
    {
        return WI.settings.showUserAgentStyles.value || style.ownerRule?.type !== WI.CSSStyleSheet.Type.UserAgent;
    }
```

this way it's easier for a future developer to search for `WI.settings.showUserAgentStyles` and see similar code everywhere it's used :)

dcrousso reviewed (approved):
nice work!

only a few adjustments :)

apologies for one of my previous suggestions

please let me know if anything is unclear

burg reviewed (approved):
Nice!

J0pan committed:
Web Inspector: Add a toggle to hide UA stylesheet rules in Elements tab
https://bugs.webkit.org/show_bug.cgi?id=302161
rdar://164265183

Reviewed by NOBODY.

This change adds a setting in the Setting > Elements tab to control whether User Agent styles are displayed.
When disabled, User Agent rules are filtered out at section creation time in the Elements > Styles panel.
Under the computed Panel the "go-to" arrows situated beside each style property are also hidden if
the style property points to a User Agent Style and the "show user agent styles" settings is disabled.

No new tests.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
(WI.ComputedStyleDetailsPanel.prototype.initialLayout):
(WI.ComputedStyleDetailsPanel.prototype._handleShowUserAgentStylesChanged):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.layout):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._shouldIncludeStyle):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._handleShowUserAgentStylesChanged):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype.update):

Canonical link: https://commits.webkit.org/304627@main
